Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp

```diff
@@ -0,0 +1,131 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
+// Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory UKRI,
+//     NScD Oak Ridge National Laboratory, European Spallation Source
+//     & Institut Laue - Langevin
+// SPDX - License - Identifier: GPL - 3.0 +
+
+#include "MantidAlgorithms/ReflectometryBackgroundSubtraction.h"
+#include "MantidAPI/CommonBinsValidator.h"
+#include "MantidKernel/EnabledWhenProperty.h"
+
+namespace Mantid {
+namespace Algorithms {
+
+/// Algorithm's name for identification. @see Algorithm::name
+const std::string ReflectometryBackgroundSubtraction::name() const {
+  return "ReflectometryBackgroundSubtraction";
+}
+
+/// Algorithm's version for identification. @see Algorithm::version
+int ReflectometryBackgroundSubtraction::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string ReflectometryBackgroundSubtraction::category() const {
+  return "Reflectometry;Reflectometry\\ISIS";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string ReflectometryBackgroundSubtraction::summary() const {
+  return "";
+}
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(ReflectometryBackgroundSubtraction)
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void ReflectometryBackgroundSubtraction::init() {
+
+  // Input workspace
+  declareProperty(make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input,
+                      boost::make_shared<CommonBinsValidator>()),
+                  "An input workspace.");
+
+  initOffspecBgdProperties();
+  initSNSBgdProperties();
+  initILLBgdProperties();
+
+  // Offspec output workspace
+  declareProperty(make_unique<WorkspaceProperty<>>("OutputWorkspaceOffspec", "",
+                                                   Direction::Output),
+                  "A Workspace with the background removed calculated with a "
+                  "basic offspec script.");
+
+
+  // SNS output workspace
+  declareProperty(make_unique<WorkspaceProperty<>>("OutputWorkspaceSNS", "",
+                                                   Direction::Output),
+                  "A Workspace with the background removed calculated using "
+                  "the SNS liquids calculation.");
+
+  // ILL output workspace
+  declareProperty(make_unique<WorkspaceProperty<>>("OutputWorkspaceILL", "",
```

Comment:
Would it be better to have a single output workspace and a drop-down box to select which correction is required?